hash.go

```diff
@@ -0,0 +1,112 @@
+package bn256
+
+// Hash implements a hashing function into the G1 group. It uses Fouque-Tibouchi encoding described here
+// https://tools.ietf.org/pdf/draft-irtf-cfrg-hash-to-curve-04.pdf
+func Hash(msg []byte) *G1 {
+	return mapToCurve(hashToBase(msg))
+}
+
+func mapToCurve(t *gfP) *G1 {
+	// calculate w = (s * t)/(1 + B + t^2)
+	w := &gfP{}
+
+	// s is a square root of -3
+	s := &gfP{0x236e675956be783b, 0x053957e6f379ab64, 0xe60789a768f4a5c4, 0x04f8979dd8bad754}
+
+	t2 := &gfP{}
+	gfpMul(t2, t, t)
+	gfpAdd(w, curveB, t2)
+	gfpAdd(w, w, &one)
+	w.Invert(w)
+
+	gfpMul(w, w, s)
+	gfpMul(w, w, t)
+
+	e := legendre(t)
+	cp := &curvePoint{z: one, t: one}
+
+	// calculate x1 = ((-1 + s) / 2) - t * w
+	tw := &gfP{}
+	gfpMul(tw, t, w)
+	x1 := &gfP{}
+	gfpAdd(x1, s, newGFp(-1))
+	half := newGFp(2)
+	half.Invert(half)
+	gfpMul(x1, x1, half)
+	gfpSub(x1, x1, tw)
+
+	// check if y=x1^3+3 is a square
+	y := &gfP{}
+	y.Set(x1)
+	gfpMul(y, x1, x1)
+	gfpMul(y, y, x1)
+	gfpAdd(y, y, curveB)
+	if legendre(y) == 1 {
+		cp.x = *x1
+		y.Sqrt(y)
+		if e == -1 {
+			gfpNeg(y, y)
+		}
+		cp.y = *y
+		return &G1{cp}
+	}
+
+	// calculate x2 = -1 - x1
+	x2 := newGFp(-1)
+	gfpSub(x2, x2, x1)
+
+	// check if y=x2^3+3 is a square
+	y.Set(x2)
+	gfpMul(y, x2, x2)
+	gfpMul(y, y, x2)
+	gfpAdd(y, y, curveB)
+	if legendre(y) == 1 {
+		cp.x = *x2
+		y.Sqrt(y)
+		if e == -1 {
+			gfpNeg(y, y)
+		}
+		cp.y = *y
+		return &G1{cp}
+	}
+
+	// calculate x3 = 1 + (1/ww)
+	x3 := &gfP{}
+	gfpMul(x3, w, w)
+	w.Invert(w)
```

Comment:
The calculation of inverses can be merged into a single function.
```tex 
z = 1/(xy)
invX = z * y
invY = z * x
```

Reply:
done.

Reply:
done.